src/edu/csus/ecs/pc2/clics/API202306/CLICSContestAccess.java

```diff
@@ -0,0 +1,168 @@
+            cap.add(API_CAPABILITY_PROXY_SUBMIT);
+        }
+        if(ClarificationService.isProxySubmitClarificationAllowed(sc)) {
+            cap.add(API_CAPABILITY_PROXY_CLAR);
+        }
+        if(SubmissionService.isAdminSubmitAllowed(sc)) {
+            cap.add(API_CAPABILITY_ADMIN_SUBMIT);
+        }
+        if(ClarificationService.isAdminSubmitClarificationAllowed(sc)) {
+            cap.add(API_CAPABILITY_ADMIN_CLAR);
+        }
+        if(AwardService.isAwardsUpdateAllowed(sc)) {
+            cap.add(API_CAPABILITY_AWARDS_UPDATE);
+        }
+        capabilities = cap.toArray(new String[0]);
+
+        /*
+         * As per the spec, here are the possible endpoints.
+         * These should match the class names in this package, eg. Contest -> ContestService, Run -> RunService, etc.
+         */
+        String [] serviceNames = { "Contest", "JudgementType", "Language", "Problem", "Group", "Organization",
+                "Team", "Person", "Account", "State", "Submission", "Judgement", "Run", "Clarification",
+                "Scoreboard", "EventFeed", "Award", "Commentary", "Webhooks"
+        };
+
+        ArrayList<CLICSEndpoint> epList = new ArrayList<CLICSEndpoint>();
+
+        for (String service: serviceNames) {
+            if(!invokeServiceEndpointAccessMethod(epList, service, sc)) {
+                if(Utilities.isDebugMode()) {
+                    System.err.println("No endpoint access for service " + service);
+                }
+                controller.getLog().log(Log.WARNING, "No endpoint access for service " + service);
+            }
+        }
+        endpoints = epList.toArray(new CLICSEndpoint[0]);
+    }
+
+    /**
+     * Create a fully qualified class name for the supplied web service endpoint.
+     * eg. Group would create: edu.csus.ecs.pc2.clics.API202306.GroupService
+     * Then call its getEndpointProperties(SecurityContext) static method to see what properties the
+     * service supports for the supplied SecurityContext (user).
+     * Note: If Java reflection makes you queasy, please stop here and just skip reading the code in this method.
+     *
+     * @param epList List to add the CLICSEndpointObject to, if one was found
+     * @param serviceName The service name only (excluding the trailing "Service")
+     * @param sc User's security context
+     * @return true if this endpoint is accessible by the user in the SecurityContext
+     */
+    private boolean invokeServiceEndpointAccessMethod(ArrayList<CLICSEndpoint> epList, String serviceName, SecurityContext sc) {
+
+        Class c = this.getClass();
+        String className = c.getPackage().getName() + "." + serviceName + "Service";
+        try {
+            Class<?> newClass = Class.forName(className);
+            try {
+                Method m = newClass.getMethod(ENDPOINT_ACCESS_METHOD, IInternalContest.class, SecurityContext.class);
+                if(m != null) {
+                    // arguments to EP access metho
```

Comment:
Trivial typo:  metho should be method...

Reply:
Fixed in next push.